{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script permite aceder ao NCBI e a extracção das informações relativas ao gene a estudar (FDX1), guardando-as num ficheiro do tipo genbank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"beatrizmariarm@gmail.com\"\n",
    "handle = Entrez.efetch(db=\"nucleotide\",rettype=\"gb\",retmode=\"text\", id=\"NC_000011.10\", seq_start=\"1\",seq_stop=\"34937\")\n",
    "seq_record=SeqIO.read(handle,\"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(seq_record,\"sequencefdx1.gb\",\"genbank\")\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REGISTO NCBI\n",
      "\n",
      "ID: NC_000011.10\n",
      "Name: NC_000011\n",
      "Description: Homo sapiens chromosome 11, GRCh38.p13 Primary Assembly\n",
      "Database cross-references: BioProject:PRJNA168, Assembly:GCF_000001405.39\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=09-DEC-2019\n",
      "/accessions=['NC_000011']\n",
      "/sequence_version=10\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Human chromosome 11 DNA sequence and analysis including novel gene identification', ...), Reference(title='Finishing the euchromatic sequence of the human genome', ...), Reference(title='Initial sequencing and analysis of the human genome', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CM000673.2.\n",
      "On Feb 3, 2014 this sequence version replaced NC_000011.9.\n",
      "Assembly Name: GRCh38.p13 Primary Assembly\n",
      "The DNA sequence is composed of genomic sequence, primarily\n",
      "finished clones that were sequenced as part of the Human Genome\n",
      "Project. PCR products and WGS shotgun sequence have been added\n",
      "where necessary to fill gaps or correct errors. All such additions\n",
      "are manually curated by GRC staff. For more information see:\n",
      "https://genomereference.org.\n",
      "Release 109.20191205\n",
      "pipeline\n",
      "RefSeq model\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Updated annotation'), ('Annotation Name', 'Homo sapiens Updated Annotation'), ('Annotation Version', '109.20191205'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation'), ('Annotation Software Version', '8.3'), ('Annotation Method', 'Best-placed RefSeq; propagated'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', IUPACAmbiguousDNA())\n"
     ]
    }
   ],
   "source": [
    "record=SeqIO.read(\"sequencefdx1.gb\",\"genbank\")\n",
    "print(\"REGISTO NCBI\" + \"\\n\")\n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Da mesma forma, o script seguinte permite aceder ao NCBI e a extracção das informações relativas à proteína do FDX1, guardando-as num ficheiro do tipo genbank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle=Entrez.efetch(db=\"protein\",rettype=\"gb\",retmode=\"text\",id=\"NP_004100.1\", seq_start=\"1\", seq_stop=\"184\")\n",
    "seq_record=SeqIO.read(handle,\"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(seq_record, 'proteinafdx1.gb', \"genbank\")\n",
    "handle.close ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REGISTO NCBI\n",
      "\n",
      "ID: NP_004100.1\n",
      "Name: NP_004100\n",
      "Description: adrenodoxin, mitochondrial precursor [Homo sapiens]\n",
      "Number of features: 13\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=06-JUL-2019\n",
      "/accessions=['NP_004100']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Active Site Structures of CYP11A1 in the Presence of Its Physiological Substrates and Alterations upon Binding of Adrenodoxin', ...), Reference(title='Molecular Recognition in Mitochondrial Cytochromes P450 That Catalyze the Terminal Steps of Corticosteroid Biosynthesis', ...), Reference(title='Human Mitochondrial Ferredoxin 1 (FDX1) and Ferredoxin 2 (FDX2) Both Bind Cysteine Desulfurase and Donate Electrons for Iron-Sulfur Cluster Biosynthesis', ...), Reference(title='Association between CCDC132, FDX1 and TNFSF13 gene polymorphisms and the risk of IgA nephropathy', ...), Reference(title='Genetic polymorphisms in TNFSF13 and FDX1 are associated with IgA nephropathy in the Han Chinese population', ...), Reference(title='Identification by site-directed mutagenesis of two lysine residues in cholesterol side chain cleavage cytochrome P450 that are essential for adrenodoxin binding', ...), Reference(title='Site-specific mutations in human ferredoxin that affect binding to ferredoxin reductase and cytochrome P450scc', ...), Reference(title='1H NMR spectra of vertebrate [2Fe-2S] ferredoxins. Hyperfine resonances suggest different electron delocalization patterns from plant ferredoxins', ...), Reference(title='Regional mapping of genes encoding human steroidogenic enzymes: P450scc to 15q23-q24, adrenodoxin to 11q22; adrenodoxin reductase to 17q24-q25; and P450c17 to 10q24-q25', ...), Reference(title='Structure, sequence, chromosomal location, and evolution of the human ferredoxin gene family', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from DB459133.1, BC010284.1,\n",
      "CN333249.1 and AA505582.1.\n",
      "Summary: This gene encodes a small iron-sulfur protein that\n",
      "transfers electrons from NADPH through ferredoxin reductase to\n",
      "mitochondrial cytochrome P450, involved in steroid, vitamin D, and\n",
      "bile acid metabolism. Pseudogenes of this functional gene are found\n",
      "on chromosomes 20 and 21. [provided by RefSeq, Aug 2011].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "[ECO:0000332]\n",
      "SAMEA1965299, SAMEA1966682\n",
      "[ECO:0000348]\n",
      "ENSP00000260270.2\n",
      "protein-coding transcript\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'BC010284.1, SRR1803617.201753.1'), ('RNAseq introns', 'single sample supports all introns')])), ('RefSeq-Attributes', OrderedDict([('gene product(s) localized to mito.', 'reported by MitoCarta'), ('MANE Ensembl match', 'ENST00000260270.3/'), ('RefSeq Select criteria', 'based on single')]))])\n",
      "Seq('MAAAGGARLLRAASAVLGGPAGRWLHHAGSRAGSSGLLRNRGPGGSAEASRSLS...KTS', IUPACProtein())\n"
     ]
    }
   ],
   "source": [
    "record=SeqIO.read(\"proteinafdx1.gb\",\"genbank\")\n",
    "print (\"REGISTO NCBI\" + \"\\n\")\n",
    "print (record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script dá-nos a quantidade total de features do gene em estudo, identificando os do tipo CDS. Como também, fornece informação quanto aos qualifiers e faz a tradução do gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index 0, ID = NC_000011.10, lenght 34937, with 1 features\n",
      "ID: NC_000011.10\n",
      "Name: NC_000011\n",
      "Description: Homo sapiens chromosome 11, GRCh38.p13 Primary Assembly\n",
      "Database cross-references: BioProject:PRJNA168, Assembly:GCF_000001405.39\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=09-DEC-2019\n",
      "/accessions=['NC_000011']\n",
      "/sequence_version=10\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Human chromosome 11 DNA sequence and analysis including novel gene identification', ...), Reference(title='Finishing the euchromatic sequence of the human genome', ...), Reference(title='Initial sequencing and analysis of the human genome', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CM000673.2.\n",
      "On Feb 3, 2014 this sequence version replaced NC_000011.9.\n",
      "Assembly Name: GRCh38.p13 Primary Assembly\n",
      "The DNA sequence is composed of genomic sequence, primarily\n",
      "finished clones that were sequenced as part of the Human Genome\n",
      "Project. PCR products and WGS shotgun sequence have been added\n",
      "where necessary to fill gaps or correct errors. All such additions\n",
      "are manually curated by GRC staff. For more information see:\n",
      "https://genomereference.org.\n",
      "Release 109.20191205\n",
      "pipeline\n",
      "RefSeq model\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Updated annotation'), ('Annotation Name', 'Homo sapiens Updated Annotation'), ('Annotation Version', '109.20191205'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation'), ('Annotation Software Version', '8.3'), ('Annotation Method', 'Best-placed RefSeq; propagated'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN', IUPACAmbiguousDNA())\n",
      "Features CDS\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "for index, record in enumerate(SeqIO.parse(open(\"sequencefdx1.gb\"),\"genbank\")):\n",
    "        print (\"index %i, ID = %s, lenght %i, with %i features\"\\\n",
    "        % (index, record.id, len(record.seq), len(record.features)))\n",
    "        print (record)\n",
    "\n",
    "\n",
    "featcds=[]\n",
    "for i in range(len(record.features)):\n",
    "    if record.features[i].type == \"CDS\":\n",
    "        featcds.append(i)\n",
    "\n",
    "print(\"Features CDS\")\n",
    "print(len(featcds))\n",
    "\n",
    "for k in featcds:\n",
    "    print (record.features[k].qualifiers[\"product\"])\n",
    "    print(\"\")\n",
    "    print(\"Tradução:\",record.features[k].qualifiers[\"translation\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script dá-nos a quantidade total de features da proteína do FDX1, identificando os do tipo CDS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index 0, ID = NP_004100.1, lenght 184, with 13 features\n",
      "ID: NP_004100.1\n",
      "Name: NP_004100\n",
      "Description: adrenodoxin, mitochondrial precursor [Homo sapiens]\n",
      "Number of features: 13\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=06-JUL-2019\n",
      "/accessions=['NP_004100']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Active Site Structures of CYP11A1 in the Presence of Its Physiological Substrates and Alterations upon Binding of Adrenodoxin', ...), Reference(title='Molecular Recognition in Mitochondrial Cytochromes P450 That Catalyze the Terminal Steps of Corticosteroid Biosynthesis', ...), Reference(title='Human Mitochondrial Ferredoxin 1 (FDX1) and Ferredoxin 2 (FDX2) Both Bind Cysteine Desulfurase and Donate Electrons for Iron-Sulfur Cluster Biosynthesis', ...), Reference(title='Association between CCDC132, FDX1 and TNFSF13 gene polymorphisms and the risk of IgA nephropathy', ...), Reference(title='Genetic polymorphisms in TNFSF13 and FDX1 are associated with IgA nephropathy in the Han Chinese population', ...), Reference(title='Identification by site-directed mutagenesis of two lysine residues in cholesterol side chain cleavage cytochrome P450 that are essential for adrenodoxin binding', ...), Reference(title='Site-specific mutations in human ferredoxin that affect binding to ferredoxin reductase and cytochrome P450scc', ...), Reference(title='1H NMR spectra of vertebrate [2Fe-2S] ferredoxins. Hyperfine resonances suggest different electron delocalization patterns from plant ferredoxins', ...), Reference(title='Regional mapping of genes encoding human steroidogenic enzymes: P450scc to 15q23-q24, adrenodoxin to 11q22; adrenodoxin reductase to 17q24-q25; and P450c17 to 10q24-q25', ...), Reference(title='Structure, sequence, chromosomal location, and evolution of the human ferredoxin gene family', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from DB459133.1, BC010284.1,\n",
      "CN333249.1 and AA505582.1.\n",
      "Summary: This gene encodes a small iron-sulfur protein that\n",
      "transfers electrons from NADPH through ferredoxin reductase to\n",
      "mitochondrial cytochrome P450, involved in steroid, vitamin D, and\n",
      "bile acid metabolism. Pseudogenes of this functional gene are found\n",
      "on chromosomes 20 and 21. [provided by RefSeq, Aug 2011].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "[ECO:0000332]\n",
      "SAMEA1965299, SAMEA1966682\n",
      "[ECO:0000348]\n",
      "ENSP00000260270.2\n",
      "protein-coding transcript\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'BC010284.1, SRR1803617.201753.1'), ('RNAseq introns', 'single sample supports all introns')])), ('RefSeq-Attributes', OrderedDict([('gene product(s) localized to mito.', 'reported by MitoCarta'), ('MANE Ensembl match', 'ENST00000260270.3/'), ('RefSeq Select criteria', 'based on single')]))])\n",
      "Seq('MAAAGGARLLRAASAVLGGPAGRWLHHAGSRAGSSGLLRNRGPGGSAEASRSLS...KTS', IUPACProtein())\n",
      "Features CDS\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "for index, record in enumerate(SeqIO.parse(open(\"proteinafdx1.gb\"),\"genbank\")):\n",
    "        print (\"index %i, ID = %s, lenght %i, with %i features\"\\\n",
    "        % (index, record.id, len(record.seq), len(record.features)))\n",
    "        print (record)\n",
    "\n",
    "\n",
    "featcds=[]\n",
    "for i in range(len(record.features)):\n",
    "    if record.features[i].type == \"CDS\":\n",
    "        featcds.append(i)\n",
    "\n",
    "print(\"Features CDS\")\n",
    "print(len(featcds))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função imprimir a localização dos features do gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Análise de features\n",
      "\n",
      "Features gene:\n",
      "0\n",
      "Locais:\n",
      "[]\n",
      "\n",
      "\n",
      "Features CDS:\n",
      "0\n",
      "Locais:\n",
      "[]\n",
      "\n",
      "Outras features:\n",
      "1\n",
      "Locais: \n",
      "[0]\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.read(\"sequencefdx1.gb\", \"genbank\")\n",
    "\n",
    "print ( \"Análise de features\")\n",
    "print (\"\")\n",
    "featuregene=[]\n",
    "featurecds = []\n",
    "outros=[]\n",
    "for i in range(len(record.features)): \n",
    "    if record.features[i].type == \"CDS\": \n",
    "        featurecds.append(i)\n",
    "    elif (record.features[i].type == \"gene\"):\n",
    "        featuregene.append(i)\n",
    "    else:\n",
    "        outros.append(i)\n",
    "    i=+1\n",
    "\n",
    "\n",
    "\n",
    "CDS_geneID =[]\n",
    "CDS_proteinID =[]\n",
    "\n",
    "for i in record.features:\n",
    "    if i.type == \"CDS\":\n",
    "         CDS_geneID.append(i.qualifiers[\"db_xref\"][1].strip(\"GeneID:\"))\n",
    "\n",
    "\n",
    "\n",
    "print(\"Features gene:\")\n",
    "print(len(featuregene))\n",
    "print(\"Locais:\")\n",
    "print(featuregene)\n",
    "print(\"\")\n",
    "print(\"\\n\" + \"Features CDS:\")\n",
    "print(len(featurecds))\n",
    "print(\"Locais:\")\n",
    "print(featurecds)\n",
    "print(\"\")\n",
    "print(\"Outras features:\")\n",
    "print(len(outros))\n",
    "print(\"Locais: \")\n",
    "print(outros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função imprimir a localização dos features da proteína do FDX1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Análise de features\n",
      "\n",
      "Features gene:\n",
      "0\n",
      "Locais:\n",
      "[]\n",
      "\n",
      "\n",
      "Features CDS:\n",
      "1\n",
      "Locais:\n",
      "[12]\n",
      "\n",
      "Outras features:\n",
      "12\n",
      "Locais: \n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.read(\"proteinafdx1.gb\", \"genbank\")\n",
    "\n",
    "print ( \"Análise de features\")\n",
    "print (\"\")\n",
    "featuregene=[]\n",
    "featurecds = []\n",
    "outros=[]\n",
    "for i in range(len(record.features)): \n",
    "    if record.features[i].type == \"CDS\": \n",
    "        featurecds.append(i)\n",
    "    elif (record.features[i].type == \"gene\"):\n",
    "        featuregene.append(i)\n",
    "    else:\n",
    "        outros.append(i)\n",
    "    i=+1\n",
    "\n",
    "\n",
    "\n",
    "CDS_geneID =[]\n",
    "CDS_proteinID =[]\n",
    "\n",
    "for i in record.features:\n",
    "    if i.type == \"CDS\":\n",
    "         CDS_geneID.append(i.qualifiers[\"db_xref\"][1].strip(\"GeneID:\"))\n",
    "\n",
    "\n",
    "\n",
    "print(\"Features gene:\")\n",
    "print(len(featuregene))\n",
    "print(\"Locais:\")\n",
    "print(featuregene)\n",
    "print(\"\")\n",
    "print(\"\\n\" + \"Features CDS:\")\n",
    "print(len(featurecds))\n",
    "print(\"Locais:\")\n",
    "print(featurecds)\n",
    "print(\"\")\n",
    "print(\"Outras features:\")\n",
    "print(len(outros))\n",
    "print(\"Locais: \")\n",
    "print(outros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função criar um ficheiro, chamado \"pesquisa.txt\", no qual apresenta uma pesquisa no pubmed do texto \"FDX1, melanoma\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "\n",
    "Entrez.email = \"beatrizmariarm@gmail.com\"\n",
    "handle = Entrez.egquery(term = \"FDX1, melanoma\")\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "\n",
    "for row in record[\"eGQueryResult\"]:\n",
    "    if row[\"DbName\"]==\"pubmed\":\n",
    "        y = row[\"Count\"]\n",
    "        \n",
    "\n",
    "handle = Entrez.esearch(db = \"pubmed\", term = \"FDX1, melanoma\", retmax=y)\n",
    "record = Entrez.read(handle)\n",
    "idlist = record[\"IdList\"]\n",
    "\n",
    "\n",
    "handle = Entrez.efetch(db=\"pubmed\", id=idlist, rettype=\"medline\", retmode=\"text\")\n",
    "records = list(Medline.parse(handle))\n",
    "record_results = open('pesquisa.txt', 'w')\n",
    "for record in records:\n",
    "    titulo=('Titulo: ', record.get('TI', '?'))\n",
    "    autores=('Autores: ', record.get('AU', '?'))\n",
    "    fonte=('Fonte: ', record.get('SO', '?'))\n",
    "    \n",
    "    record_results.write(str(titulo))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(str(autores))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(str(fonte))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(\"\\n\")\n",
    "record_results.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O ficheiro \"proteinfdx1.fasta\" contém a sequência de aminoácidos da proteína do FDX1. Este ficheiro é necessário para o correr os próximos dois scripts\n",
    "#O primeiro script dá informações sobre o tamanho da sequência e corre o Blast para realizar uma pesquisa com recurso ao SwissProt por forma a encontrar sequências similares, criando um ficheiro .xml. Seguidamente o programa realiza o parsing com recurso ao ficheiro criado anteriormente, sendo analisados vários parâmetros como a base de dados utilizada, a matriz de substituição e os parâmetros que definem os espaçamentos. São identidicados accesion numbers, e-values e os comprimentos dos alinhamentos que foram devolvidos.\n",
    "#O procedimento é repetido para o organismo \"Homo Sapiens\" e os resultados são novamente guardados em ficheiro .xml. É identificado o alinhamento com menor e-value e identificado o registo da Uniprot da sequência e o número de HSP's, analisando posteriormente o mesmo para todos os alinhamentos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "184\n",
      "PARAMETERS:\n",
      "Database: swissprot_v5\n",
      "Matrix: BLOSUM62\n",
      "Gap penalties:  (11, 1)\n",
      "number hits:  50\n",
      "E-values and length of alignments:\n",
      "P10109 - 7.0724e-133 length:184\n",
      "P00258 - 4.64323e-99 length:185\n",
      "P00257 - 1.16224e-95 length:180\n",
      "P24483 - 6.76732e-87 length:188\n",
      "P46656 - 3.15147e-84 length:188\n",
      "P13216 - 1.89148e-81 length:134\n",
      "P29330 - 4.76428e-80 length:124\n",
      "Q8SZA8 - 1.6575e-44 length:149\n",
      "Q92J08 - 8.74886e-34 length:110\n",
      "Q4UKL2 - 4.27689e-33 length:103\n",
      "Q9AKM6 - 4.81995e-33 length:103\n",
      "Q9AKC4 - 3.1738e-32 length:103\n",
      "Q9AKH1 - 4.18765e-32 length:103\n",
      "Q9ZDW6 - 2.51249e-31 length:103\n",
      "Q9M0V0 - 1.01528e-29 length:105\n",
      "Q8S904 - 8.61954e-29 length:118\n",
      "Q1RJ69 - 4.27455e-26 length:102\n",
      "Q12184 - 2.40675e-25 length:112\n",
      "Q5FWQ0 - 1.8191e-23 length:107\n",
      "Q8SV19 - 2.15934e-22 length:110\n",
      "Q9CPW2 - 4.27181e-22 length:170\n",
      "P37193 - 9.46335e-22 length:109\n",
      "Q05B51 - 1.06396e-21 length:163\n",
      "Q08C57 - 2.83633e-21 length:109\n",
      "Q6P4F2 - 3.54207e-21 length:177\n",
      "Q10361 - 4.55227e-20 length:85\n",
      "P57661 - 1.96225e-11 length:92\n",
      "P37098 - 2.89104e-11 length:106\n",
      "P44428 - 4.6633e-11 length:93\n",
      "X5CWH9 - 5.71122e-11 length:99\n",
      "P0A9R4 - 8.71658e-11 length:103\n",
      "P43493 - 1.41515e-10 length:104\n",
      "Q5S3I4 - 3.58096e-10 length:102\n",
      "Q51383 - 3.90314e-10 length:90\n",
      "P80306 - 1.444e-09 length:101\n",
      "O51882 - 2.18141e-09 length:70\n",
      "P33007 - 4.09041e-09 length:102\n",
      "P00259 - 2.88971e-08 length:102\n",
      "Q89A15 - 2.35731e-07 length:71\n",
      "X5CFH4 - 2.53135e-07 length:103\n",
      "D5IGG4 - 8.61385e-06 length:90\n",
      "G2IN77 - 0.00181537 length:99\n",
      "P0A4I9 - 1.01547 length:62\n",
      "Q13087 - 1.62202 length:41\n",
      "O33662 - 1.62418 length:62\n",
      "Q0P5N6 - 2.09488 length:122\n",
      "Q1BYS6 - 2.57296 length:82\n",
      "A0K4X6 - 2.57296 length:82\n",
      "B1JWH3 - 2.59659 length:82\n",
      "A5FL33 - 2.64694 length:65\n",
      "Accession: P10109\n",
      "Hit id: sp|P10109.1|\n",
      "Definition: RecName: Full=Adrenodoxin, mitochondrial; AltName: Full=Adrenal ferredoxin; AltName: Full=Ferredoxin-1; AltName: Full=Hepatoredoxin; Flags: Precursor [Homo sapiens]\n",
      "Number of HSPs:  1\n",
      "E-value:  4.49176e-134\n",
      "Length:  184\n",
      "Identities:  184\n",
      "Query start:  1\n",
      "Sbjct start:  1\n",
      "MAAAGGARLLRAASAVLGGPAGRWLHHAGSRAGSSGLLRNRGPGGSAEASRSLSVSARARSSSEDKITVHFINRDGETLTTKGKVGDSLL\n",
      "MAAAGGARLLRAASAVLGGPAGRWLHHAGSRAGSSGLLRNRGPGGSAEASRSLSVSARARSSSEDKITVHFINRDGETLTTKGKVGDSLL\n",
      "MAAAGGARLLRAASAVLGGPAGRWLHHAGSRAGSSGLLRNRGPGGSAEASRSLSVSARARSSSEDKITVHFINRDGETLTTKGKVGDSLL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW \n",
    "from Bio import SeqIO\n",
    "\n",
    "#abre o ficheiro e vê o tamanho da sequência\n",
    "seqrecord = SeqIO.read(open(\"proteinfdx1.fasta\"), format=\"fasta\") \n",
    "print (len(seqrecord.seq))\n",
    "  \n",
    "#pesquisa sequências similares através do blast, com recurso à swissprot;\n",
    "#o resultado obtido é guardado num ficheiro xml\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"))\n",
    "\n",
    "save_file = open(\"blast-fdx1.xml\", \"w\")\n",
    "save_file.write(result_handle.read()) \n",
    "save_file.close() \n",
    "result_handle.close()\n",
    "\n",
    "#parsing\n",
    "result_handle = open(\"blast-fdx1.xml\")\n",
    "record = NCBIXML.read(result_handle)\n",
    "\n",
    "\n",
    "print (\"PARAMETERS:\")\n",
    "print (\"Database: \" + record.database)\n",
    "print (\"Matrix: \" + record.matrix)\n",
    "print (\"Gap penalties: \", record.gap_penalties)\n",
    "\n",
    "#características dos alinhamentos\n",
    "nhits = len(record.alignments) \n",
    "print (\"number hits: \", nhits)\n",
    "\n",
    "res = []\n",
    "for alignment in record.alignments:\n",
    "    evalue = alignment.hsps[0].expect\n",
    "    accession = alignment.accession\n",
    "    leng = alignment.hsps[0].align_length\n",
    "    res.append(accession + \" - \" + str(evalue) + \" length:\" + str(leng) )\n",
    "\n",
    "print(\"E-values and length of alignments:\")\n",
    "for s in res: print(s)\n",
    "\n",
    "#blast para Homo Sapiens\n",
    "result_handle2 = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"), entrez_query = \"Homo sapiens[organism]\" )\n",
    "\n",
    "save_file = open(\"blast-fdx1.xml\", \"w\")\n",
    "save_file.write(result_handle2.read()) \n",
    "save_file.close() \n",
    "result_handle2.close()\n",
    "\n",
    "#análise dos resultados\n",
    "result_handle2 = open(\"blast-fdx1.xml\")\n",
    "record2 = NCBIXML.read(result_handle2)\n",
    "\n",
    "first_alignment = record2.alignments[0]\n",
    "\n",
    "print (\"Accession: \" + first_alignment.accession)\n",
    "print (\"Hit id: \" + first_alignment.hit_id)\n",
    "print (\"Definition: \" + first_alignment.hit_def)\n",
    "\n",
    "print (\"Number of HSPs: \", len(first_alignment.hsps))\n",
    "\n",
    "for hsp in first_alignment.hsps:\n",
    "    print (\"E-value: \", hsp.expect)\n",
    "    print (\"Length: \", hsp.align_length)\n",
    "    print (\"Identities: \", hsp.identities)\n",
    "    print (\"Query start: \", hsp.query_start)\n",
    "    print (\"Sbjct start: \", hsp.sbjct_start)\n",
    "    print (hsp.query[0:90])\n",
    "    print (hsp.match[0:90])\n",
    "    print (hsp.sbjct[0:90])\n",
    "    print (\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O último script guarda em ficheiro fasta as sequências homólogas obtidas para poderem servir de input em programas como o MEGAX, de construção filogenética."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW \n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "seqrecord = SeqIO.read(open(\"proteinfdx1.fasta\"), format=\"fasta\")\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"))\n",
    "\n",
    "save_file = open(\"blast-fdx1.xml\", \"w\")\n",
    "save_file.write(result_handle.read()) \n",
    "save_file.close() \n",
    "result_handle.close()\n",
    "\n",
    "result_handle = open(\"blast-fdx1.xml\")\n",
    "record = NCBIXML.read(result_handle)\n",
    "\n",
    "nhits = len(record.alignments) \n",
    "print (\"number hits: \", nhits)\n",
    "\n",
    "Entrez.email = 'beatrizmariarm@gmail.com'\n",
    "\n",
    "res = []\n",
    "for alignment in record.alignments:\n",
    "    evalue = alignment.hsps[0].expect\n",
    "    accession = alignment.accession\n",
    "    leng = alignment.hsps[0].align_length\n",
    "    \n",
    "    #guardar em ficheiro fasta as sequências homólogas obtidas\n",
    "    handle = Entrez.efetch(db=\"protein\", id=accession, rettype=\"fasta\", retmode=\"text\")\n",
    "    seq_record = SeqIO.read(handle, \"fasta\")\n",
    "    SeqIO.write(seq_record, accession + '.fasta', \"fasta\") \n",
    "    handle.close()\n",
    "    \n",
    "    res.append(accession + \" - \" + str(evalue) + \" length:\" + str(leng) )\n",
    "\n",
    "print(\"E-values and length of alignments:\")\n",
    "for s in res: print(s)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
