{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SLC25A28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script permite aceder ao NCBI e a extracção das informações relativas ao gene a estudar (FDX1), guardando-as num ficheiro do tipo genbank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"beatrizmariarm@gmail.com\"\n",
    "handle = Entrez.efetch(db=\"nucleotide\",rettype=\"gb\",retmode=\"text\", id=\"NC_000010.11\", seq_start=\"1\",seq_stop=\"48986\")\n",
    "seq_record=SeqIO.read(handle,\"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(seq_record,\"sequenceslc.gb\",\"genbank\")\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REGISTO NCBI\n",
      "\n",
      "ID: NC_000010.11\n",
      "Name: NC_000010\n",
      "Description: Homo sapiens chromosome 10, GRCh38.p13 Primary Assembly\n",
      "Database cross-references: BioProject:PRJNA168, Assembly:GCF_000001405.39\n",
      "Number of features: 14\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=09-DEC-2019\n",
      "/accessions=['NC_000010']\n",
      "/sequence_version=11\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Finishing the euchromatic sequence of the human genome', ...), Reference(title='The DNA sequence and comparative analysis of human chromosome 10', ...), Reference(title='Initial sequencing and analysis of the human genome', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CM000672.2.\n",
      "On Feb 3, 2014 this sequence version replaced NC_000010.10.\n",
      "Assembly Name: GRCh38.p13 Primary Assembly\n",
      "The DNA sequence is composed of genomic sequence, primarily\n",
      "finished clones that were sequenced as part of the Human Genome\n",
      "Project. PCR products and WGS shotgun sequence have been added\n",
      "where necessary to fill gaps or correct errors. All such additions\n",
      "are manually curated by GRC staff. For more information see:\n",
      "https://genomereference.org.\n",
      "Release 109.20191205\n",
      "pipeline\n",
      "RefSeq model\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Updated annotation'), ('Annotation Name', 'Homo sapiens Updated Annotation'), ('Annotation Version', '109.20191205'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation'), ('Annotation Software Version', '8.3'), ('Annotation Method', 'Best-placed RefSeq; propagated'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...CAC', IUPACAmbiguousDNA())\n"
     ]
    }
   ],
   "source": [
    "record=SeqIO.read(\"sequenceslc.gb\",\"genbank\")\n",
    "print(\"REGISTO NCBI\" + \"\\n\")\n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Da mesma forma, o script seguinte permite aceder ao NCBI e a extracção das informações relativas à proteína do SLC25A28, guardando-as num ficheiro do tipo genbank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle=Entrez.efetch(db=\"protein\",rettype=\"gb\",retmode=\"text\",id=\"NP_112489.3\", seq_start=\"1\", seq_stop=\"364\")\n",
    "seq_record=SeqIO.read(handle,\"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(seq_record, 'proteinaslc.gb', \"genbank\")\n",
    "handle.close ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REGISTO NCBI\n",
      "\n",
      "ID: NP_112489.3\n",
      "Name: NP_112489\n",
      "Description: mitoferrin-2 [Homo sapiens]\n",
      "Number of features: 15\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=02-AUG-2019\n",
      "/accessions=['NP_112489']\n",
      "/sequence_version=3\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Identification of susceptibility loci for colorectal cancer in a genome-wide meta-analysis', ...), Reference(title='The mitochondrial transporter family SLC25: identification, properties and physiopathology', ...), Reference(title='Genetic determinants of plasma beta(2)-glycoprotein I levels: a genome-wide association study in extended pedigrees from Spain', ...), Reference(title='Mitoferrin-2-dependent mitochondrial iron uptake sensitizes human head and neck squamous carcinoma cells to photodynamic therapy', ...), Reference(title=\"A genome-wide scan of Ashkenazi Jewish Crohn's disease suggests novel susceptibility loci\", ...), Reference(title='A germline-specific class of small RNAs binds mammalian Piwi proteins', ...), Reference(title='A scan of chromosome 10 identifies a novel locus showing strong association with late-onset Alzheimer disease', ...), Reference(title='The DNA sequence and comparative analysis of human chromosome 10', ...), Reference(title='Expressed sequence tag analysis of adult human iris for the NEIBank Project: steroid-response factors and similarities with retinal pigment epithelium', ...), Reference(title='Characterization of a novel human putative mitochondrial transporter homologous to the yeast mitochondrial RNA splicing proteins 3 and 4', ...)]\n",
      "/comment=VALIDATED REFSEQ: This record has undergone validation or\n",
      "preliminary review. The reference sequence was derived from\n",
      "CD672723.1 and BC047312.1.\n",
      "On Jun 21, 2007 this sequence version replaced NP_112489.2.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "SAMEA2150385, SAMN01820691\n",
      "[ECO:0000348]\n",
      "ENSP00000359526.4\n",
      "expression, longest protein\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'BC047312.1, AF327402.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns')])), ('RefSeq-Attributes', OrderedDict([('gene product(s) localized to mito.', 'reported by MitoCarta'), ('MANE Ensembl match', 'ENST00000370495.6/'), ('RefSeq Select criteria', 'based on conservation,')]))])\n",
      "Seq('MELEGRGAGGVAGGPAAGPGRSPGESALLDGWLQRGVGRGAGGGEAGACRPPVR...AGK', IUPACProtein())\n"
     ]
    }
   ],
   "source": [
    "record=SeqIO.read(\"proteinaslc.gb\",\"genbank\")\n",
    "print (\"REGISTO NCBI\" + \"\\n\")\n",
    "print (record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script dá-nos a quantidade total de features do gene em estudo, identificando os do tipo CDS. Como também, fornece informação quanto aos qualifiers e faz a tradução do gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index 0, ID = NC_000010.11, lenght 48986, with 14 features\n",
      "ID: NC_000010.11\n",
      "Name: NC_000010\n",
      "Description: Homo sapiens chromosome 10, GRCh38.p13 Primary Assembly\n",
      "Database cross-references: BioProject:PRJNA168, Assembly:GCF_000001405.39\n",
      "Number of features: 14\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=09-DEC-2019\n",
      "/accessions=['NC_000010']\n",
      "/sequence_version=11\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Finishing the euchromatic sequence of the human genome', ...), Reference(title='The DNA sequence and comparative analysis of human chromosome 10', ...), Reference(title='Initial sequencing and analysis of the human genome', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CM000672.2.\n",
      "On Feb 3, 2014 this sequence version replaced NC_000010.10.\n",
      "Assembly Name: GRCh38.p13 Primary Assembly\n",
      "The DNA sequence is composed of genomic sequence, primarily\n",
      "finished clones that were sequenced as part of the Human Genome\n",
      "Project. PCR products and WGS shotgun sequence have been added\n",
      "where necessary to fill gaps or correct errors. All such additions\n",
      "are manually curated by GRC staff. For more information see:\n",
      "https://genomereference.org.\n",
      "Release 109.20191205\n",
      "pipeline\n",
      "RefSeq model\n",
      "/structured_comment=OrderedDict([('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Status', 'Updated annotation'), ('Annotation Name', 'Homo sapiens Updated Annotation'), ('Annotation Version', '109.20191205'), ('Annotation Pipeline', 'NCBI eukaryotic genome annotation'), ('Annotation Software Version', '8.3'), ('Annotation Method', 'Best-placed RefSeq; propagated'), ('Features Annotated', 'Gene; mRNA; CDS; ncRNA')]))])\n",
      "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...CAC', IUPACAmbiguousDNA())\n",
      "Features CDS\n",
      "5\n",
      "['tubulin beta-8 chain']\n",
      "\n",
      "Tradução: ['MREIVLTQIGQCGNQIGAKFWEVISDEHAIDSAGTYHGDSHLQLERINVYYNEASGGRYVPRAVLVDLEPGTMDSVRSGPFGQVFRPDNFIFGQCGAGNNWAKGHYTEGAELMESVMDVVRKEAESCDCLQGFQLTHSLGGGTGSGMGTLLLSKIREEYPDRIINTFSILPSPKVSDTVVEPYNATLSVHQLIENADETFCIDNEALYDICSKTLKLPTPTYGDLNHLVSATMSGVTTCLRFPGQLNADLRKLAVNMVPFPRLHFFMPGFAPLTSRGSQQYRALTVAELTQQMFDAKNMMAACDPRHGRYLTAAAIFRGRMPMREVDEQMFNIQDKNSSYFADWLPNNVKTAVCDIPPRGLKMSATFIGNNTAIQELFKRVSEQFTAMFRRKAFLHWYTGEGMDEMEFTEAESNMNDLVSEYQQYQDATAEEEEDEEYAEEEVA']\n",
      "['tubulin beta-8 chain isoform X3']\n",
      "\n",
      "Tradução: ['MDSVRSGPFGQVFRPDNFIFGQCGAGNNWAKGHYTEGAELMESVMDVVRKEAESCDCLQGFQLTHSLGGGTGSGMGTLLLSKIREEYPDRIINTFSILPSPKVSDTVVEPYNATLSVHQLIENADETFCIDNEALYDICSKTLKLPTPTYGDLNHLVSATMSGVTTCLRFPGQLNADLRKLAVNMVPFPRLHFFMPGFAPLTSRGSQQYRALTVAELTQQMFDAKNMMAACDPRHGRYLTAAAIFRGRMPMREVDEQMFNIQDKNSSYFADWLPNNVKTAVCDIPPRGLKMSATFIGNNTAIQELFKRVSEQFTAMFRRKAFLHWYTGEGMDEMEFTEAESNMNDLVSEYQQYQDATAEEEEDEEYAEEEVA']\n",
      "['tubulin beta-8 chain isoform X2']\n",
      "\n",
      "Tradução: ['MVVTTDDRIPGRRVTGCAVLQVSDTVVEPYNATLSVHQLIENADETFCIDNEALYDICSKTLKLPTPTYGDLNHLVSATMSGVTTCLRFPGQLNADLRKLAVNMVPFPRLHFFMPGFAPLTSRGSQQYRALTVAELTQQMFDAKNMMAACDPRHGRYLTAAAIFRGRMPMREVDEQMFNIQDKNSSYFADWLPNNVKTAVCDIPPRGLKMSATFIGNNTAIQELFKRVSEQFTAMFRRKAFLHWYTGEGMDEMEFTEAESNMNDLVSEYQQYQDATAEEEEDEEYAEEEVA']\n",
      "['tubulin beta-8 chain isoform X1']\n",
      "\n",
      "Tradução: ['MESVMDVVRKEAESCDCLQGFQLTHSLGGGTGSGMGTLLLSKIREEYPDRIINTFSILPSPKVSDTVVEPYNATLSVHQLIENADETFCIDNEALYDICSKTLKLPTPTYGDLNHLVSATMSGVTTCLRFPGQLNADLRKLAVNMVPFPRLHFFMPGFAPLTSRGSQQYRALTVAELTQQMFDAKNMMAACDPRHGRYLTAAAIFRGRMPMREVDEQMFNIQDKNSSYFADWLPNNVKTAVCDIPPRGLKMSATFIGNNTAIQELFKRVSEQFTAMFRRKAFLHWYTGEGMDEMEFTEAESNMNDLVSEYQQYQDATAEEEEDEEYAEEEVA']\n",
      "['tubulin beta-8 chain isoform X1']\n",
      "\n",
      "Tradução: ['MESVMDVVRKEAESCDCLQGFQLTHSLGGGTGSGMGTLLLSKIREEYPDRIINTFSILPSPKVSDTVVEPYNATLSVHQLIENADETFCIDNEALYDICSKTLKLPTPTYGDLNHLVSATMSGVTTCLRFPGQLNADLRKLAVNMVPFPRLHFFMPGFAPLTSRGSQQYRALTVAELTQQMFDAKNMMAACDPRHGRYLTAAAIFRGRMPMREVDEQMFNIQDKNSSYFADWLPNNVKTAVCDIPPRGLKMSATFIGNNTAIQELFKRVSEQFTAMFRRKAFLHWYTGEGMDEMEFTEAESNMNDLVSEYQQYQDATAEEEEDEEYAEEEVA']\n"
     ]
    }
   ],
   "source": [
    "for index, record in enumerate(SeqIO.parse(open(\"sequenceslc.gb\"),\"genbank\")):\n",
    "        print (\"index %i, ID = %s, lenght %i, with %i features\"\\\n",
    "        % (index, record.id, len(record.seq), len(record.features)))\n",
    "        print (record)\n",
    "\n",
    "\n",
    "featcds=[]\n",
    "for i in range(len(record.features)):\n",
    "    if record.features[i].type == \"CDS\":\n",
    "        featcds.append(i)\n",
    "\n",
    "print(\"Features CDS\")\n",
    "print(len(featcds))\n",
    "\n",
    "for k in featcds:\n",
    "    print (record.features[k].qualifiers[\"product\"])\n",
    "    print(\"\")\n",
    "    print(\"Tradução:\",record.features[k].qualifiers[\"translation\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script dá-nos a quantidade total de features da proteína do SLC25A28, identificando os do tipo CDS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index 0, ID = NP_112489.3, lenght 364, with 15 features\n",
      "ID: NP_112489.3\n",
      "Name: NP_112489\n",
      "Description: mitoferrin-2 [Homo sapiens]\n",
      "Number of features: 15\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=02-AUG-2019\n",
      "/accessions=['NP_112489']\n",
      "/sequence_version=3\n",
      "/keywords=['RefSeq', 'MANE Select']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='Identification of susceptibility loci for colorectal cancer in a genome-wide meta-analysis', ...), Reference(title='The mitochondrial transporter family SLC25: identification, properties and physiopathology', ...), Reference(title='Genetic determinants of plasma beta(2)-glycoprotein I levels: a genome-wide association study in extended pedigrees from Spain', ...), Reference(title='Mitoferrin-2-dependent mitochondrial iron uptake sensitizes human head and neck squamous carcinoma cells to photodynamic therapy', ...), Reference(title=\"A genome-wide scan of Ashkenazi Jewish Crohn's disease suggests novel susceptibility loci\", ...), Reference(title='A germline-specific class of small RNAs binds mammalian Piwi proteins', ...), Reference(title='A scan of chromosome 10 identifies a novel locus showing strong association with late-onset Alzheimer disease', ...), Reference(title='The DNA sequence and comparative analysis of human chromosome 10', ...), Reference(title='Expressed sequence tag analysis of adult human iris for the NEIBank Project: steroid-response factors and similarities with retinal pigment epithelium', ...), Reference(title='Characterization of a novel human putative mitochondrial transporter homologous to the yeast mitochondrial RNA splicing proteins 3 and 4', ...)]\n",
      "/comment=VALIDATED REFSEQ: This record has undergone validation or\n",
      "preliminary review. The reference sequence was derived from\n",
      "CD672723.1 and BC047312.1.\n",
      "On Jun 21, 2007 this sequence version replaced NP_112489.2.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "SAMEA2150385, SAMN01820691\n",
      "[ECO:0000348]\n",
      "ENSP00000359526.4\n",
      "expression, longest protein\n",
      "/structured_comment=OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'BC047312.1, AF327402.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns')])), ('RefSeq-Attributes', OrderedDict([('gene product(s) localized to mito.', 'reported by MitoCarta'), ('MANE Ensembl match', 'ENST00000370495.6/'), ('RefSeq Select criteria', 'based on conservation,')]))])\n",
      "Seq('MELEGRGAGGVAGGPAAGPGRSPGESALLDGWLQRGVGRGAGGGEAGACRPPVR...AGK', IUPACProtein())\n",
      "Features CDS\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "for index, record in enumerate(SeqIO.parse(open(\"proteinaslc.gb\"),\"genbank\")):\n",
    "        print (\"index %i, ID = %s, lenght %i, with %i features\"\\\n",
    "        % (index, record.id, len(record.seq), len(record.features)))\n",
    "        print (record)\n",
    "\n",
    "\n",
    "featcds=[]\n",
    "for i in range(len(record.features)):\n",
    "    if record.features[i].type == \"CDS\":\n",
    "        featcds.append(i)\n",
    "\n",
    "print(\"Features CDS\")\n",
    "print(len(featcds))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função imprimir a localização dos features do gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Análise de features\n",
      "\n",
      "Features gene:\n",
      "2\n",
      "Locais:\n",
      "[1, 3]\n",
      "\n",
      "\n",
      "Features CDS:\n",
      "5\n",
      "Locais:\n",
      "[9, 10, 11, 12, 13]\n",
      "\n",
      "Outras features:\n",
      "7\n",
      "Locais: \n",
      "[0, 2, 4, 5, 6, 7, 8]\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.read(\"sequenceslc.gb\", \"genbank\")\n",
    "\n",
    "print ( \"Análise de features\")\n",
    "print (\"\")\n",
    "featuregene=[]\n",
    "featurecds = []\n",
    "outros=[]\n",
    "for i in range(len(record.features)): \n",
    "    if record.features[i].type == \"CDS\": \n",
    "        featurecds.append(i)\n",
    "    elif (record.features[i].type == \"gene\"):\n",
    "        featuregene.append(i)\n",
    "    else:\n",
    "        outros.append(i)\n",
    "    i=+1\n",
    "\n",
    "\n",
    "\n",
    "CDS_geneID =[]\n",
    "CDS_proteinID =[]\n",
    "\n",
    "for i in record.features:\n",
    "    if i.type == \"CDS\":\n",
    "         CDS_geneID.append(i.qualifiers[\"db_xref\"][1].strip(\"GeneID:\"))\n",
    "\n",
    "\n",
    "\n",
    "print(\"Features gene:\")\n",
    "print(len(featuregene))\n",
    "print(\"Locais:\")\n",
    "print(featuregene)\n",
    "print(\"\")\n",
    "print(\"\\n\" + \"Features CDS:\")\n",
    "print(len(featurecds))\n",
    "print(\"Locais:\")\n",
    "print(featurecds)\n",
    "print(\"\")\n",
    "print(\"Outras features:\")\n",
    "print(len(outros))\n",
    "print(\"Locais: \")\n",
    "print(outros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função imprimir a localização dos features da proteína do SLC25A28."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Análise de features\n",
      "\n",
      "Features gene:\n",
      "0\n",
      "Locais:\n",
      "[]\n",
      "\n",
      "\n",
      "Features CDS:\n",
      "1\n",
      "Locais:\n",
      "[14]\n",
      "\n",
      "Outras features:\n",
      "14\n",
      "Locais: \n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]\n"
     ]
    }
   ],
   "source": [
    "record = SeqIO.read(\"proteinaslc.gb\", \"genbank\")\n",
    "\n",
    "print ( \"Análise de features\")\n",
    "print (\"\")\n",
    "featuregene=[]\n",
    "featurecds = []\n",
    "outros=[]\n",
    "for i in range(len(record.features)): \n",
    "    if record.features[i].type == \"CDS\": \n",
    "        featurecds.append(i)\n",
    "    elif (record.features[i].type == \"gene\"):\n",
    "        featuregene.append(i)\n",
    "    else:\n",
    "        outros.append(i)\n",
    "    i=+1\n",
    "\n",
    "\n",
    "\n",
    "CDS_geneID =[]\n",
    "CDS_proteinID =[]\n",
    "\n",
    "for i in record.features:\n",
    "    if i.type == \"CDS\":\n",
    "         CDS_geneID.append(i.qualifiers[\"db_xref\"][1].strip(\"GeneID:\"))\n",
    "\n",
    "\n",
    "\n",
    "print(\"Features gene:\")\n",
    "print(len(featuregene))\n",
    "print(\"Locais:\")\n",
    "print(featuregene)\n",
    "print(\"\")\n",
    "print(\"\\n\" + \"Features CDS:\")\n",
    "print(len(featurecds))\n",
    "print(\"Locais:\")\n",
    "print(featurecds)\n",
    "print(\"\")\n",
    "print(\"Outras features:\")\n",
    "print(len(outros))\n",
    "print(\"Locais: \")\n",
    "print(outros)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O script tem como principal função criar um ficheiro, chamado \"pesquisa.txt\", no qual apresenta uma pesquisa no pubmed do texto \"SLC25A28, melanoma\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "\n",
    "Entrez.email = \"beatrizmariarm@gmail.com\"\n",
    "handle = Entrez.egquery(term = \"SLC25A28, melanoma\")\n",
    "record = Entrez.read(handle)\n",
    "\n",
    "\n",
    "for row in record[\"eGQueryResult\"]:\n",
    "    if row[\"DbName\"]==\"pubmed\":\n",
    "        y = row[\"Count\"]\n",
    "        \n",
    "\n",
    "handle = Entrez.esearch(db = \"pubmed\", term = \"SLC25A28, melanoma\", retmax=y)\n",
    "record = Entrez.read(handle)\n",
    "idlist = record[\"IdList\"]\n",
    "\n",
    "\n",
    "handle = Entrez.efetch(db=\"pubmed\", id=idlist, rettype=\"medline\", retmode=\"text\")\n",
    "records = list(Medline.parse(handle))\n",
    "record_results = open('pesquisa.txt', 'w')\n",
    "for record in records:\n",
    "    titulo=('Titulo: ', record.get('TI', '?'))\n",
    "    autores=('Autores: ', record.get('AU', '?'))\n",
    "    fonte=('Fonte: ', record.get('SO', '?'))\n",
    "    \n",
    "    record_results.write(str(titulo))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(str(autores))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(str(fonte))\n",
    "    record_results.write(\"\\n\")\n",
    "    record_results.write(\"\\n\")\n",
    "record_results.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O ficheiro \"sequenceslc.fasta\" contém a sequência de aminoácidos da proteína do FDX1. Este ficheiro é necessário para o correr os próximos dois scripts\n",
    "#O primeiro script dá informações sobre o tamanho da sequência e corre o Blast para realizar uma pesquisa com recurso ao SwissProt por forma a encontrar sequências similares, criando um ficheiro .xml. Seguidamente o programa realiza o parsing com recurso ao ficheiro criado anteriormente, sendo analisados vários parâmetros como a base de dados utilizada, a matriz de substituição e os parâmetros que definem os espaçamentos. São identidicados accesion numbers, e-values e os comprimentos dos alinhamentos que foram devolvidos.\n",
    "#O procedimento é repetido para o organismo \"Homo Sapiens\" e os resultados são novamente guardados em ficheiro .xml. É identificado o alinhamento com menor e-value e identificado o registo da Uniprot da sequência e o número de HSP's, analisando posteriormente o mesmo para todos os alinhamentos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "364\n",
      "PARAMETERS:\n",
      "Database: swissprot_v5\n",
      "Matrix: BLOSUM62\n",
      "Gap penalties:  (11, 1)\n",
      "number hits:  50\n",
      "E-values and length of alignments:\n",
      "Q96A46 - 0.0 length:364\n",
      "Q8R0Z5 - 0.0 length:364\n",
      "Q7T292 - 2.10139e-179 length:343\n",
      "Q287T7 - 5.16666e-163 length:307\n",
      "Q920G8 - 2.87912e-159 length:329\n",
      "Q66H23 - 5.54953e-159 length:329\n",
      "Q9NYZ2 - 7.37485e-159 length:302\n",
      "Q23125 - 5.06401e-98 length:300\n",
      "Q9VAY3 - 7.05203e-97 length:292\n",
      "Q620A6 - 7.12672e-91 length:301\n",
      "O14281 - 2.34554e-75 length:301\n",
      "Q6GLJ0 - 6.65833e-71 length:117\n",
      "Q68F18 - 2.6808e-69 length:172\n",
      "P23500 - 3.05959e-68 length:295\n",
      "P10566 - 7.53295e-68 length:302\n",
      "Q55DY8 - 2.30156e-61 length:299\n",
      "Q3ZBJ8 - 8.81714e-52 length:113\n",
      "F4HT41 - 5.44039e-32 length:286\n",
      "Q54S10 - 8.92244e-31 length:337\n",
      "Q55E45 - 1.0174e-30 length:295\n",
      "Q03829 - 2.3104e-30 length:303\n",
      "Q86AV5 - 2.6422e-29 length:285\n",
      "Q6DHC3 - 4.41977e-28 length:340\n",
      "Q94AG6 - 3.0789e-27 length:295\n",
      "Q99297 - 4.94509e-27 length:292\n",
      "P0CAT2 - 9.20839e-27 length:279\n",
      "Q3TZX3 - 3.83704e-26 length:319\n",
      "Q9BSK2 - 1.07096e-25 length:320\n",
      "Q6P316 - 2.2663e-25 length:339\n",
      "Q09461 - 2.39518e-25 length:331\n",
      "Q0VCH6 - 2.53808e-25 length:329\n",
      "Q08CI8 - 2.55756e-25 length:278\n",
      "Q8TBP6 - 2.56366e-25 length:323\n",
      "Q1LZB3 - 5.58134e-25 length:315\n",
      "Q6P036 - 1.67396e-24 length:315\n",
      "Q75AH6 - 7.15666e-24 length:293\n",
      "Q8BGP6 - 1.74233e-23 length:339\n",
      "Q9FHX2 - 4.64979e-23 length:285\n",
      "Q6DFK2 - 6.5749e-23 length:329\n",
      "Q7SXW0 - 1.0102e-22 length:255\n",
      "Q9BZJ4 - 2.78297e-22 length:361\n",
      "Q12482 - 4.5682e-22 length:293\n",
      "Q6DJ08 - 4.6797e-22 length:278\n",
      "P38921 - 4.85398e-22 length:300\n",
      "Q96DW6 - 8.35235e-22 length:302\n",
      "Q6NYZ6 - 1.71587e-21 length:284\n",
      "Q9P6N6 - 2.37385e-21 length:297\n",
      "Q9UJS0 - 2.4746e-21 length:279\n",
      "Q9QXX4 - 2.89968e-21 length:282\n",
      "Q54RB9 - 4.25986e-21 length:281\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW \n",
    "from Bio import SeqIO\n",
    "\n",
    "#abre o ficheiro e vê o tamanho da sequência\n",
    "seqrecord = SeqIO.read(open(\"sequenceslc.fasta\"), format=\"fasta\") \n",
    "print (len(seqrecord.seq))\n",
    "  \n",
    "#pesquisa sequências similares através do blast, com recurso à swissprot;\n",
    "#o resultado obtido é guardado num ficheiro xml\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"))\n",
    "\n",
    "save_file = open(\"blast-slc.xml\", \"w\")\n",
    "save_file.write(result_handle.read()) \n",
    "save_file.close() \n",
    "result_handle.close()\n",
    "\n",
    "#parsing\n",
    "result_handle = open(\"blast-slc.xml\")\n",
    "record = NCBIXML.read(result_handle)\n",
    "\n",
    "\n",
    "print (\"PARAMETERS:\")\n",
    "print (\"Database: \" + record.database)\n",
    "print (\"Matrix: \" + record.matrix)\n",
    "print (\"Gap penalties: \", record.gap_penalties)\n",
    "\n",
    "#características dos alinhamentos\n",
    "nhits = len(record.alignments) \n",
    "print (\"number hits: \", nhits)\n",
    "\n",
    "res = []\n",
    "for alignment in record.alignments:\n",
    "    evalue = alignment.hsps[0].expect\n",
    "    accession = alignment.accession\n",
    "    leng = alignment.hsps[0].align_length\n",
    "    res.append(accession + \" - \" + str(evalue) + \" length:\" + str(leng) )\n",
    "\n",
    "print(\"E-values and length of alignments:\")\n",
    "for s in res: print(s)\n",
    "\n",
    "#blast para Homo Sapiens\n",
    "result_handle2 = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"), entrez_query = \"Homo sapiens[organism]\" )\n",
    "\n",
    "save_file = open(\"blast-slc.xml\", \"w\")\n",
    "save_file.write(result_handle2.read()) \n",
    "save_file.close() \n",
    "result_handle2.close()\n",
    "\n",
    "#análise dos resultados\n",
    "result_handle2 = open(\"blast-slc.xml\")\n",
    "record2 = NCBIXML.read(result_handle2)\n",
    "\n",
    "first_alignment = record2.alignments[0]\n",
    "\n",
    "print (\"Accession: \" + first_alignment.accession)\n",
    "print (\"Hit id: \" + first_alignment.hit_id)\n",
    "print (\"Definition: \" + first_alignment.hit_def)\n",
    "\n",
    "print (\"Number of HSPs: \", len(first_alignment.hsps))\n",
    "\n",
    "for hsp in first_alignment.hsps:\n",
    "    print (\"E-value: \", hsp.expect)\n",
    "    print (\"Length: \", hsp.align_length)\n",
    "    print (\"Identities: \", hsp.identities)\n",
    "    print (\"Query start: \", hsp.query_start)\n",
    "    print (\"Sbjct start: \", hsp.sbjct_start)\n",
    "    print (hsp.query[0:90])\n",
    "    print (hsp.match[0:90])\n",
    "    print (hsp.sbjct[0:90])\n",
    "    print (\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#O último script guarda em ficheiro fasta as sequências homólogas obtidas para poderem servir de input em programas como o MEGAX, de construção filogenética."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW \n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "seqrecord = SeqIO.read(open(\"sequenceslc.fasta\"), format=\"fasta\")\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"swissprot\", seqrecord.format(\"fasta\"))\n",
    "\n",
    "save_file = open(\"blast-slc.xml\", \"w\")\n",
    "save_file.write(result_handle.read()) \n",
    "save_file.close() \n",
    "result_handle.close()\n",
    "\n",
    "result_handle = open(\"blast-slc.xml\")\n",
    "record = NCBIXML.read(result_handle)\n",
    "\n",
    "nhits = len(record.alignments) \n",
    "print (\"number hits: \", nhits)\n",
    "\n",
    "Entrez.email = 'beatrizmariarm@gmail.com'\n",
    "\n",
    "res = []\n",
    "for alignment in record.alignments:\n",
    "    evalue = alignment.hsps[0].expect\n",
    "    accession = alignment.accession\n",
    "    leng = alignment.hsps[0].align_length\n",
    "    \n",
    "    #guardar em ficheiro fasta as sequências homólogas obtidas\n",
    "    handle = Entrez.efetch(db=\"protein\", id=accession, rettype=\"fasta\", retmode=\"text\")\n",
    "    seq_record = SeqIO.read(handle, \"fasta\")\n",
    "    SeqIO.write(seq_record, accession + '.fasta', \"fasta\") \n",
    "    handle.close()\n",
    "    \n",
    "    res.append(accession + \" - \" + str(evalue) + \" length:\" + str(leng) )\n",
    "\n",
    "print(\"E-values and length of alignments:\")\n",
    "for s in res: print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
